### PR TITLE
YD-512 Log multiple overlapping activities

### PR DESCRIPTION
**Pull request by aristotelos**

### CODE REVIEW

**Comment by yonabuild** [comment]
Build triggered. sha1 is original commit.

**Comment by yonabuild** [comment]
Build started sha1 is original commit.

**Comment by yonabuild** [comment]
Build finished.

**Comment by yonabuild** [comment]
Build triggered. sha1 is merged.

**Comment by yonabuild** [comment]
Build started sha1 is merged.

**Comment by yonabuild** [comment]
Build finished.

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java [hunk 1]
```diff
@@ -246,9 +252,40 @@ private void addOrUpdateDayTruncatedActivity(UserAnonymizedEntityHolder userAnon
...
+		// (that indicates that there is something wrong in the app activity bookkeeping).
+		// Network activity (having payload.application == null) also is not expected to exist and overlap, as network activity
+		// has a very short time span.
+		// So log if there are multiple overlaps.
+
+		if (overlappingOfSameApp.isEmpty())
+		{
+			return Optional.empty();
+		}
+		else if (overlappingOfSameApp.size() == 1)
```
Comment (Bert-R) [inline analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: Else isn't necessary here.

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java [hunk 2]
```diff
@@ -246,9 +252,40 @@ private void addOrUpdateDayTruncatedActivity(UserAnonymizedEntityHolder userAnon
...
+		{
+			return Optional.of(overlappingOfSameApp.get(0));
+		}
+		else
```
Comment (Bert-R) [inline analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: Same here

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java [hunk 3]
```diff
@@ -246,9 +252,40 @@ private void addOrUpdateDayTruncatedActivity(UserAnonymizedEntityHolder userAnon
...
+		{
+			String relatedTo = payload.application.isPresent()
+					? MessageFormat.format("same app ''{0}'' activity", payload.application.get()) : "network activity";
```
Comment (Bert-R) [inline analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: This now results in messages like this:
* ``Multiple overlapping activities of same app 'abc' activity found.``
* ``Multiple overlapping activities of network activity found.``

Can't you change it this way?
* ``Multiple overlapping app activities of 'abc' found.``
* ``Multiple overlapping network activities found.``

That would be more clear to me.

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java [hunk 4]
```diff
@@ -246,9 +252,40 @@ private void addOrUpdateDayTruncatedActivity(UserAnonymizedEntityHolder userAnon
...
+			String overlappingActivities = overlappingOfSameApp.stream()
+					.map(overlappingActivity -> overlappingActivity.toString()).collect(Collectors.joining(", "));
+			logger.warn(
+					"Multiple overlapping activities of {0} found. The payload has start time {1} and end time {2}. The overlapping activities are: {3}. The day activity ID is {4} and the activity category ID is {5}.",
```
Comment (Bert-R) [inline analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: I think the understandability would improve if the last part of the message was the collection of overlapping activities. That would require something like this:
```
"Multiple overlapping activities of {0} found. The payload has start time {1} and end time {2}. The day activity ID is {4} and the activity category ID is {5}. The overlapping activities are: {3}."
```

Note that SLF4J does not use sequence numbers for the arguments but expects the parameters to come in the right order, so the arguments need to be reordered and the numbers removed.

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java (hunk 1 above)
Comment (aristotelos) [reply analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: Removed

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java (hunk 2 above)
Comment (aristotelos) [reply analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: Removed

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java (hunk 3 above)
Comment (aristotelos) [reply analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: Good point, changed

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java (hunk 4 above)
Comment (aristotelos) [reply analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: Changed and fixed

**Comment by yonabuild** [comment]
Build triggered. sha1 is merged.

**Comment by yonabuild** [comment]
Build started sha1 is merged.

File: core/src/main/java/nu/yona/server/analysis/entities/Activity.java [hunk 5]
```diff
@@ -137,4 +138,12 @@ public DeviceAnonymized getDeviceAnonymized()
 	{
 		return deviceAnonymized;
 	}
+
+	@Override
+	public String toString()
+	{
+		return MessageFormat.format(
+				"activity from {0} to {1} (timezone {2}) of activity category with id {3} and app ''{4}'' and device anonymized with id {5} (activity id {6})",
+				startTime, endTime, timeZone, activityCategory.getId(), app, deviceAnonymized.getId(), getId());
```
[the remark this answers is not in the document]
Comment (aristotelos) [reply core/src/main/java/nu/yona/server/analysis/entities/Activity.java]: OK, will add a check

**Comment by yonabuild** [comment]
Build finished.

File: core/src/main/java/nu/yona/server/analysis/entities/Activity.java [hunk 6]
```diff
@@ -137,4 +138,12 @@ public DeviceAnonymized getDeviceAnonymized()
 	{
 		return deviceAnonymized;
 	}
+
+	@Override
+	public String toString()
+	{
+		return MessageFormat.format(
+				"activity from {0} to {1} (timezone {2}) of activity category with id {3} and app ''{4}'' and device anonymized with id {5} (activity id {6})",
```
[the remark this answers is not in the document]
Comment (aristotelos) [reply core/src/main/java/nu/yona/server/analysis/entities/Activity.java]: Yes it does.

**Comment by yonabuild** [comment]
Build triggered. sha1 is merged.

**Comment by yonabuild** [comment]
Build started sha1 is merged.

**Comment by yonabuild** [comment]
Build finished.

**Comment by yonabuild** [comment]
Build triggered. sha1 is merged.

**Comment by yonabuild** [comment]
Build started sha1 is merged.

**Comment by yonabuild** [comment]
Build finished.

**Pull request merged**